crates/cxx-qt-gen/src/naming/type_names.rs

```diff
@@ -21,9 +23,88 @@ use crate::parser::{cxxqtdata::ParsedCxxQtData, qenum::ParsedQEnum, qobject::Par
 ///
 /// This is used by the generator phase to find types by their identifier, such that they can be
 /// fully qualified in Rust and C++.
-#[derive(Default, Debug)]
+#[derive(Debug)]
 pub struct TypeNames {
     names: BTreeMap<Ident, Name>,
+    // Currently, there is only a single case for which duplicate type names are allowed.
+    // That is if a type is declared as a shared type, and declared as external.
+    // So store the names of the types that are declared as shared and external, so we can
+    // check for duplicates in all other cases.
+    extern_types: BTreeSet<Ident>,
+    shared_types: BTreeSet<Ident>,
+}
+
+impl Default for TypeNames {
+    fn default() -> Self {
+        let trivial = |rust_name, namespace: &str, cxx_name: &str| Name {
+            rust: format_ident!("{}", rust_name),
+            cxx: Some(cxx_name.to_owned()),
+            module: None,
+            namespace: Some(namespace.to_owned()),
+        };
+
+        // By default, CXX supports a set of types that don't need to be declared:
+        // See: https://github.com/dtolnay/cxx/blob/master/syntax/atom.rs
+        // and: https://github.com/dtolnay/cxx/blob/a6e1cd1e8d9d6df20e88e7443963dc4c5c8c4875/gen/src/write.rs#L1311
+        let cxx_types = [
```

Comment:
this stuff would be better in the separate lib?

As these were helpers that other CXX tools could use for going between CXX Rust and C++ types.

Reply:
I think the idea was to extract the whole naming module into its own library if there's interest in using it.
So I think that's alright?